modules/ml/src/main/java/org/apache/ignite/ml/trees/trainers/columnbased/ColumnDecisionTreeTrainer.java

```diff
@@ -0,0 +1,559 @@
+                        }
+                        else {
+                            rightBlock.add(right);
+                            return rightBlock.equals(k) ?
+                                Stream.of(new CacheEntryImpl<>(k, leftBlock)) :
+                                Stream.of(new CacheEntryImpl<>(k, leftBlock), new CacheEntryImpl<>(rightKey, rightBlock));
+                        }
+                    },
+                    bestRegsKeys);
+
+                X.println(">>> Update of projs cache took " + (System.currentTimeMillis() - before));
+
+                before = System.currentTimeMillis();
+
+                updateSplitCache(ind, rc, featuresCnt, ig -> i -> input.affinityKey(i, ig), uuid);
+
+                X.println(">>> Update of split cache took " + (System.currentTimeMillis() - before));
+            }
+            else {
+                X.println(">>> Best feature index: " + bestFeatureIdx + ", best infoGain " + bestInfoGain);
+                break;
+            }
+        }
+
+        int rc = regsCnt;
+
+        IgniteSupplier<Iterable<Cache.Entry<RegionKey, List<RegionProjection>>>> featZeroRegs = () -> {
+            IgniteCache<RegionKey, List<RegionProjection>> projsCache = ProjectionsCache.getOrCreate(Ignition.localIgnite());
+
+            return () -> IntStream.range(0, (rc - 1) / BLOCK_SIZE + 1).
+                mapToObj(rBIdx -> ProjectionsCache.key(0, rBIdx, input.affinityKey(0, Ignition.localIgnite()), uuid)).
+                map(k -> (Cache.Entry<RegionKey, List<RegionProjection>>)new CacheEntryImpl<>(k, projsCache.localPeek(k))).iterator();
+        };
+
+        Map<Integer, Double> vals = CacheUtils.reduce(prjsCache.getName(), ignite,
+            (TrainingContext ctx, Cache.Entry<RegionKey, List<RegionProjection>> e, Map<Integer, Double> m) -> {
+                int regBlockIdx = e.getKey().regionBlockIndex();
+
+                if (e.getValue() != null) {
+                    for (int i = 0; i < e.getValue().size(); i++) {
+                        int regIdx = regBlockIdx * BLOCK_SIZE + i;
+                        RegionProjection reg = e.getValue().get(i);
+
+                        Double res = regCalc.apply(Arrays.stream(reg.sampleIndexes()).mapToDouble(s -> ctx.labels()[s]));
+                        m.put(regIdx, res);
+                    }
+                }
+
+                return m;
+            },
+            () -> ContextCache.getOrCreate(Ignition.localIgnite()).get(uuid),
+            featZeroRegs,
+            (infos, infos2) -> {
+                Map<Integer, Double> res = new HashMap<>();
+                res.putAll(infos);
+                res.putAll(infos2);
+                return res;
+            },
+            HashMap::new
+        );
+
+        int i = 0;
+        for (TreeTip tip : tips) {
+            tip.leafSetter.accept(new Leaf(vals.get(i)));
+            i++;
+        }
+
+        ProjectionsCache.clear(featuresCnt, rc, input::affinityKey, uuid, ignite);
```

Comment:
Please document action with these caches. Why behavior is different for different caches?